### PR TITLE
Generic slist-based free list structure (with criterion tests)

### LINKED ISSUE #164: Generic slist-based freelist support
To be used for gnix_fab_req, gnix_queue, etc.

### PR DESCRIPTION
**Pull request by sungeunchoi**
Clients should wrap these functions for a cleaner interface.

Note that the generic version is a little less appealing than a
type-specific version in a number of (obvious) ways:
- Chunks can't be zeroed with calloc()
- Structures can't be initialized as they are getting added to the list
- If a particular alignment is required of the individual elements,
  the element size must be padded by the client

There are probably others.

Closes #164 

Signed-off-by: Sung-Eun Choi sungeunchoi@users.noreply.github.com

### CODE REVIEW

**Comment by hppritcha** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/49/
Test PASSed.

**Comment by sungeunchoi** [comment]
@hppritcha @jswaro @ztiffany 

See what you guys think.  I'm not tied to any of the naming or interfaces.  See the criterion test for how a client would wrap this and use it for a particular struct.

File: prov/gni/src/gnix_freelist.c [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	assert(fl);
+
+	struct slist_entry *se = slist_remove_head(&fl->freelist);
+
+	if (!se) {
+		ret = __gnix_sfl_refill(fl, fl->refill_size);
+		if (ret != FI_SUCCESS)
+			goto err;
+		if (fl->refill_size < fl->max_refill_size) {
+			int ns = fl->refill_size *= fl->growth_factor;
```
Comment (hppritcha) [inline prov/gni/src/gnix_freelist.c:150]: does a declaration here pass the linux style checker?  not a big deal but just curious.

File: prov/gni/src/gnix_freelist.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		se = slist_remove_head(&fl->freelist);
+		if (!se) {
+			/* Can't happen unless multithreaded */
+			ret = -FI_EAGAIN;
+			goto err;
+		}
+	}
+
+	/* zero the actual element */
+	memset(((unsigned char *) se)-fl->offset, 0, fl->elem_size);
```
Comment (hppritcha) [inline prov/gni/src/gnix_freelist.c]: we may not want a memset for critical path struct allocations.

**Comment by hppritcha** [comment]
okay this looks good except for the `memset`.  Could we leave that out if compiling with debugging disabled?

File: prov/gni/src/gnix_freelist.c (hunk 1 above)
Comment (sungeunchoi) [inline prov/gni/src/gnix_freelist.c:150]: Yep it does.

File: prov/gni/src/gnix_freelist.c (hunk 2 above)
Comment (sungeunchoi) [inline prov/gni/src/gnix_freelist.c]: I'm fine with just removing it.

**Comment by hppritcha** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/53/
Test PASSed.

**Pull request merged**